netlink/conntracker.go

```diff
@@ -0,0 +1,266 @@
+// +build linux
+
+package netlink
+
+import (
+	"context"
+	golog "log"
+	"net"
+	"os"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+
+	ct "github.com/florianl/go-conntrack"
+)
+
+// Conntracker is a wrapper around go-conntracker that keeps a record of all connections in user space
+type Conntracker interface {
+	GetConntrackEntryForConn(ip string, port uint16) *IPTranslation
+	Close()
+}
+
+type connKey struct {
+	ip   string
+	port uint16
+}
+
+type realConntracker struct {
+	nfct  *ct.Nfct
+	l     *sync.Mutex
+	state map[connKey]*IPTranslation
+
+	statsTicker   *time.Ticker
+	compactTicker *time.Ticker
+	stats         struct {
+		gets               int64
+		getTimeTotal       int64
+		registers          int64
+		registersTotalTime int64
+	}
+}
+
+func NewConntracker() (Conntracker, error) {
+	nfct, err := ct.Open(&ct.Config{ReadTimeout: 10 * time.Millisecond, Logger: golog.New(os.Stdout, "go-conntrack", 0)})
+	if err != nil {
+		return nil, err
+	}
+
+	ctr := &realConntracker{
+		nfct:          nfct,
+		l:             &sync.Mutex{},
+		statsTicker:   time.NewTicker(time.Second * 10),
+		compactTicker: time.NewTicker(time.Hour),
+		state:         make(map[connKey]*IPTranslation),
+	}
+
+	// seed the state
+	sessions, err := nfct.Dump(ct.Ct, ct.CtIPv4)
+	if err != nil {
+		return nil, err
+	}
+	ctr.registerUnlocked(sessions)
+
+	sessions, err = nfct.Dump(ct.Ct, ct.CtIPv6)
+	if err != nil {
+		// this is not fatal because we've already seeded with IPv4
+		log.Errorf("Failed to dump IPv6")
+	}
+	ctr.registerUnlocked(sessions)
+
+	go ctr.emitStats()
+	go ctr.compact()
```

Comment:
Can we just do both of these things in a single routine?

Reply:
They happen at different intervals